src/Chapter3.hs

```diff
@@ -480,6 +520,67 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Building
+  = Church
+  | Library deriving (Show)
+
+data HousePeople
+  = P1
+  | P2
+  | P3
+  | P4 deriving (Show)
+
+peopleCount :: HousePeople -> Int
+peopleCount P1 = 1
+peopleCount P2 = 2
+peopleCount P3 = 3
+peopleCount P4 = 4
+
+newtype House
+  = House { housePeople :: HousePeople }
+  deriving Show
+
+data Wall = Wall deriving (Show)
+
+data City = City
+  { cityCastleName   :: String
+  , cityWalls        :: [Wall]
```

Comment:
This is the correct data type, that can represent what we have in requirements 👍🏼 
However, this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable

Reply:
Nice! I was wondering if this was possible 👍